src/hooks/paraswap/useCollateralRepaySwap.tsx

```diff
@@ -0,0 +1,188 @@
+import { BigNumberZeroDecimal, normalize } from '@aave/math-utils';
+import { OptimalRate } from 'paraswap-core';
+import { useCallback, useEffect, useMemo, useState } from 'react';
+
+import {
+  fetchExactInRate,
+  fetchExactInTxParams,
+  fetchExactOutRate,
+  fetchExactOutTxParams,
+  MESSAGE_MAP,
+  SwapData,
+  SwapTransactionParams,
+  SwapVariant,
+  UseSwapProps,
+} from './common';
+
+type UseRepayWithCollateralProps = UseSwapProps & {
+  swapVariant: SwapVariant;
+};
+
+interface UseRepayWithCollateralResponse {
+  outputAmount: string;
+  outputAmountUSD: string;
+  inputAmount: string;
+  inputAmountUSD: string;
+  loading: boolean;
+  error: string;
+  buildTxFn: () => Promise<SwapTransactionParams>;
+}
+
+export const useCollateralRepaySwap = ({
+  chainId,
+  max,
+  maxSlippage,
+  skip,
+  swapIn,
+  swapOut,
+  userAddress,
+  swapVariant,
+}: UseRepayWithCollateralProps): UseRepayWithCollateralResponse => {
+  const [inputAmount, setInputAmount] = useState<string>('0');
+  const [inputAmountUSD, setInputAmountUSD] = useState<string>('0');
+  const [outputAmount, setOutputAmount] = useState<string>('0');
+  const [outputAmountUSD, setOutputAmountUSD] = useState<string>('0');
+  const [route, setRoute] = useState<OptimalRate>();
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState('');
+
+  const swapInData = useMemo(() => {
+    const swapData: SwapData = {
+      underlyingAsset: swapIn.underlyingAsset,
+      decimals: swapIn.decimals,
+      supplyAPY: swapIn.supplyAPY,
+      amount: swapIn.amount,
+      variableBorrowAPY: swapIn.variableBorrowAPY,
+    };
+    return swapData;
+  }, [
+    swapIn.amount,
+    swapIn.decimals,
+    swapIn.supplyAPY,
+    swapIn.underlyingAsset,
+    swapIn.variableBorrowAPY,
+  ]);
+
+  const swapOutData = useMemo(() => {
+    const swapData: SwapData = {
+      underlyingAsset: swapOut.underlyingAsset,
+      decimals: swapOut.decimals,
+      supplyAPY: swapOut.supplyAPY,
+      amount: swapOut.amount,
+      variableBorrowAPY: swapOut.variableBorrowAPY,
+    };
+    return swapData;
+  }, [
+    swapOut.amount,
+    swapOut.decimals,
+    swapOut.supplyAPY,
+    swapOut.underlyingAsset,
+    swapOut.variableBorrowAPY,
+  ]);
```

Comment:
Since these are basically the same function, this could be simplified with:

```js
const createMemoizedSwapData = (src: unknown): SwapData => {
  const triggers = [
    src.amount,
    src.decimals,
    src.supplyAPY,
    src.underlyingAsset,
    src.variableBorrowAPY,
  ]
  const data: SwapData = {
    underlyingAsset: src.underlyingAsset,
    decimals: src.decimals,
    supplyAPY: src.supplyAPY,
    amount: src.amount,
    variableBorrowAPY: src.variableBorrowAPY,
  }
  return useMemo(() => data, triggers)
}

const swapInData = createMemoizedSwapData(swapIn)
const swapOutData = createMemoizedSwapData(swapOut)
```

Reply:
This won't work. You can only call `useMemo` inside a component or a hook.